packages/grafana-ui/src/components/GraphNG/utils.ts

```diff
@@ -20,7 +20,9 @@ function isVisibleBarField(f: Field) {
 
 // will mutate the DataFrame's fields' values
 function applySpanNullsThresholds(frame: DataFrame) {
-  let refField = frame.fields.find((field) => field.type === FieldType.time); // this doesnt need to be time, just any numeric/asc join field
+  let refField =
+    frame.fields.find((field) => field.type === FieldType.time) ??
+    frame.fields.find((field) => field.type === FieldType.number);
```

Comment:
This change fixes the span nulls threshold functionality (without this the panel crashes when switching to threshold mode in span nulls editor) - this will select first numeric field if no time field is found.

https://github.com/grafana/grafana/assets/22381771/89e0519c-c095-4eb1-afcf-a58566d1b321

Reply:
:grimacing: a lot of this feels too handwavy :(

`applySpanNullsThresholds(frame: DataFrame)`, should take an explicit `refField` argument since in Trend the user explicitly selects which number field to use, which is not necessarily the first one.

the same thing applies to `applyNullInsertThreshold()`, where `refFieldName` should be passed in explicitly and not be null to fall back to first numeric field.

we actually pass `dimFields` field matchers for `x` and `y` to `preparePlotFrame`. right now x is hard-coded to time field, which should override on our new `preparePlotFrameTimeless` wrapper.

in the core `preparePlotFrame()` we should actually use these matchers for all prep ops (not just outerJoin) and select the correct fields instead of relying on these fallbacks.

e.g. here:

https://github.com/grafana/grafana/blob/82f0ef6871ec42f9bf16a282acdcbe1023b10145/packages/grafana-ui/src/components/GraphNG/utils.ts#L48

https://github.com/grafana/grafana/blob/82f0ef6871ec42f9bf16a282acdcbe1023b10145/packages/grafana-ui/src/components/GraphNG/utils.ts#L51

https://github.com/grafana/grafana/blob/82f0ef6871ec42f9bf16a282acdcbe1023b10145/packages/grafana-ui/src/components/GraphNG/utils.ts#L48

https://github.com/grafana/grafana/blob/82f0ef6871ec42f9bf16a282acdcbe1023b10145/packages/grafana-ui/src/components/GraphNG/utils.ts#L105